gateway/ln-gateway/src/lightning/lnd.rs

```diff
@@ -464,77 +474,163 @@ impl ILnRpcClient for GatewayLndClient {
+                    amt_msat: send_amt_msat,
                     final_cltv_delta,
+                    fee_limit: Some(FeeLimit {
+                        limit: Some(Limit::FixedMsat(fee_limit_msat)),
+                    }),
                     cltv_limit,
-                    no_inflight_updates: false,
-                    timeout_seconds: LND_PAYMENT_TIMEOUT_SECONDS,
-                    fee_limit_msat,
+                    route_hints: route_hints_to_lnd(&invoice.route_hints),
+                    dest_features,
+                    use_mission_control: true,
                     ..Default::default()
                 })
                 .await
                 .map_err(|status| {
-                    info!("LND payment request failed for invoice {invoice:?} with {status:?}");
+                    info!("LND failed to query routes for invoice {invoice:?} with {status:?}");
                     LightningRpcError::FailedPayment {
-                        failure_reason: format!("Failed to make outgoing payment {status:?}"),
+                        failure_reason: format!("Failed to make query routes {status:?}"),
                     }
-                })?;
+                })?
+                .into_inner();
+
+            let total_amt_msat: i64 = invoice.amount.msats.try_into().map_err(|error| {
+                LightningRpcError::FailedPayment {
+                    failure_reason: format!("amount exceeds valid LND amount ranges {error:?}"),
+                }
+            })?;
+
+            let mut last_failed_reason: Option<String> = None;
+
+            let start = duration_since_epoch();
+
+            for mut route in routes.routes {
+                // skip any empty routes
+                if route.hops.is_empty() {
+                    continue;
+                }
+                // check if we've timed out
+                if duration_since_epoch() > Duration::from_secs(LND_PAYMENT_TIMEOUT_SECONDS) + start
+                {
+                    return Err(LightningRpcError::FailedPayment {
+                        failure_reason: "Timed out waiting for payment to complete".to_string(),
+                    });
+                }
 
-            debug!(
-                "LND payment request sent for invoice {invoice:?}, waiting for payment status..."
-            );
-            let mut messages = payments.into_inner();
-            loop {
-                match messages
-                    .message()
+                // add mpp record to route
+                let last = route.hops.last_mut().expect("must have at least one hop");
+
+                last.mpp_record = Some(MppRecord {
+                    payment_addr: invoice.payment_secret.to_vec(),
+                    total_amt_msat,
+                });
+
+                trace!("Attempting to send payment to route {route:?}");
+
+                debug!("LND payment request sent for invoice {invoice:?}, waiting for payment status...");
```

Comment:
Can we also log what route attempt we are on? I could see that being interesting